File: interfaces/pkcs12.js
/* pkcs12.js
 *
 * Copyright (C) 2006-2022 wolfSSL Inc.
 *
 * This file is part of wolfSSL.
 *
 * wolfSSL is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * wolfSSL is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
 */
const wolfcrypt = require( '../build/Release/wolfcrypt' )

class WolfSSL_PKCS12
{
    static PBE_MD5_DES = 0;
    static PBE_SHA1_RC4_128 = 1;
    static PBE_SHA1_DES = 2;
    static PBE_SHA1_DES3 = 3;
    static PBE_AES256_CBC = 4;
    static PBE_AES128_CBC = 5;
    static PBE_SHA1_40RC2_CBC = 6;
    static PBE_SHA1_RC4_128_SUM = 657;
    static PBE_SHA1_DES3_SUM = 659;
    static PBE_SHA1_40RC2_CBC_SUM = 662;
    static PBE_MD5_DES_SUM = 651;
    static PBE_SHA1_DES_SUM = 658;
    static PBES2_SUM = 661;
    static PBES2 = 13;
    static PBES1_MD5_DES = 3;
    static PBES1_SHA1_DES = 10;

    /**
     * Creates a new PKCS7 structure by calling sizeof_PKCS7 and wc_PKCS7_Init
     *
     * @remarks free must be called to free the ecc PKCS7 data
     */
    constructor() {
        this.pkcs12 = wolfcrypt.wc_PKCS12_new()
    }

    Create(pass, key, cert, derList, nidKey, nidCert, iter, macIter) {
        if (typeof pass == 'string') {
            pass = Buffer.from(pass)
        }

        if (!Buffer.isBuffer(pass)) {
            throw 'pass must be a string or Buffer'
        }

        if (typeof key == 'string') {
            key = Buffer.from(key)
        }

        if (!Buffer.isBuffer(key)) {
            throw 'key must be a string or Buffer'
        }

        if (typeof cert == 'string') {
            cert = Buffer.from(cert)
        }

        if (!Buffer.isBuffer(cert)) {
            throw 'cert must be a string or Buffer'
        }

        for (let i in derList) {
            if (typeof derList[i] == 'string') {
                derList[i] = Buffer.from(derList[i])
            }

            if (!Buffer.isBuffer(derList[i])) {
                throw 'derList must be an array of buffers or strings'
            }
        }

        this.pkcs12 = wolfcrypt.nodejsPKCS12Create(pass, pass.length, key,
            key.length, cert, cert.length, derList, nidKey, nidCert, iter,
            macIter)

        if (this.pkcs12 == null) {
            throw 'failed to wc_PKCS12_create'
        }
    }

    Parse(pass) {
        if (this.pkcs12 == null) {
            throw 'PKCS12 not allocated'
        }

        if (typeof pass == 'string') {
            pass = Buffer.from(pass)
        }

        if (!Buffer.isBuffer(pass)) {
            throw 'pass must be a string or Buffer'
        }

        let derList = []

        let out = wolfcrypt.nodejsPKCS12Parse(this.pkcs12, pass, derList)

        if (!out.key || !out.cert) {
            throw `failed to nodejsPKCS12Parse`
        }

        return {key: out.key, cert: out.cert, derList: derList}
    }

    DerToInternal(der) {
        if (this.pkcs12 == null) {
            throw 'PKCS12 not allocated'
        }

        if (typeof der == 'string') {
            pass = Buffer.from(der)
        }

        if (!Buffer.isBuffer(der)) {
            throw 'der must be a string or Buffer'
        }

        let ret = wolfcrypt.wc_d2i_PKCS12(der, der.length, this.pkcs12)

        if (ret != 0) {
            throw `failed to wc_d2i_PKCS12 ${ret}`
        }
    }

    InternalToDer() {
        if (this.pkcs12 == null) {
            throw 'PKCS12 not allocated'
        }

        let der = wolfcrypt.nodejsPKCS12InternalToDer(this.pkcs12)

        if (der == null) {
            throw `failed to nodejsPKCS12InternalToDer`
        }

        return der;
    }

    /**
     * Frees the data allocated by the PKCS12 structure
     *
     * @throws {Error} If PKCS12 structure is not allocated.
     */
    free() {
        if ( this.pkcs12 == null )
        {
            throw 'Pkcs12 not allocated'
        }

        wolfcrypt.wc_PKCS12_free(this.pkcs12)

        this.pkcs12 = null
    }
}

exports.WolfSSL_PKCS12 = WolfSSL_PKCS12

File: addon/wolfcrypt/pkcs12.cpp
/* pkcs12.cpp
 *
 * Copyright (C) 2006-2022 wolfSSL Inc.
 *
 * This file is part of wolfSSL.
 *
 * wolfSSL is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * wolfSSL is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
 */
#include "./h/pkcs12.h"

Napi::Value bind_wc_PKCS12_new(const Napi::CallbackInfo& info)
{
  Napi::Env env = info.Env();
  WC_PKCS12* pkcs12 = wc_PKCS12_new();
  Napi::External<WC_PKCS12> pkcs12_ext = Napi::External<WC_PKCS12>::New(env,
      pkcs12);

  return pkcs12_ext;
}

Napi::Value nodejsPKCS12Create(const Napi::CallbackInfo& info)
{
    Napi::Env env = info.Env();
    uint8_t* pass = info[0].As<Napi::Uint8Array>().Data();
    int passSz = info[1].As<Napi::Number>().Int32Value();
    uint8_t* key = info[2].As<Napi::Uint8Array>().Data();
    int keySz = info[3].As<Napi::Number>().Int32Value();
    uint8_t* cert = info[4].As<Napi::Uint8Array>().Data();
    int certSz = info[5].As<Napi::Number>().Int32Value();
    Napi::Array derList = info[6].As<Napi::Array>();
    int nidKey = info[7].As<Napi::Number>().Int32Value();
    int nidMac = info[8].As<Napi::Number>().Int32Value();
    int iter = info[9].As<Napi::Number>().Int32Value();
    int macIter = info[10].As<Napi::Number>().Int32Value();
    int i;
    WC_PKCS12* pkcs12 = NULL;
    Napi::External<WC_PKCS12> pkcs12_ext;
    WC_DerCertList* root = NULL;
    WC_DerCertList* cur = NULL;

    if (derList.Length() > 0) {
        root = (WC_DerCertList*)XMALLOC(sizeof(WC_DerCertList), NULL,
            DYNAMIC_TYPE_PKCS);
        XMEMSET(root, 0, sizeof(WC_DerCertList));
        cur = root;
    }

    for (i = 0; i < (int)derList.Length(); i++) {
        Napi::Value v = derList[i];

        if (v.IsBuffer()) {
            cur->bufferSz = (word32)v.As<Napi::Buffer<uint8_t>>().Length();
            cur->buffer = (byte*)XMALLOC(cur->bufferSz, NULL,
                DYNAMIC_TYPE_PKCS);
            XMEMCPY(cur->buffer, v.As<Napi::Uint8Array>().Data(),
                cur->bufferSz);
        }

        if (i < (int)derList.Length() - 1) {
            cur->next = (WC_DerCertList*)XMALLOC(sizeof(WC_DerCertList), NULL,
                DYNAMIC_TYPE_PKCS);
            cur = cur->next;
            XMEMSET(cur, 0, sizeof(WC_DerCertList));
        }
    }

    pkcs12 = wc_PKCS12_create((char*)pass, passSz, (char*)"friendlyName", key,
        keySz, cert, certSz, root, nidKey, nidMac, iter, macIter, 0, NULL);

    if (root != NULL) {
        wc_FreeCertList(root, NULL);
    }

    if (pkcs12 != NULL) {
        pkcs12_ext = Napi::External<WC_PKCS12>::New(env, pkcs12);
    }

    return pkcs12_ext;
}

Napi::Object nodejsPKCS12Parse(const Napi::CallbackInfo& info)
{
    Napi::Env env = info.Env();
    WC_PKCS12* pkcs12 = info[0].As<Napi::External<WC_PKCS12>>().Data();
    uint8_t* pass = info[1].As<Napi::Uint8Array>().Data();
    Napi::Array derList = info[2].As<Napi::Array>();
    Napi::Object out = Napi::Object::New(env);
    Napi::Buffer<uint8_t> key;
    Napi::Buffer<uint8_t> cert;
    int ret;
    int i = 0;
    byte* tmpKey = NULL;
    word32 tmpKeySz;
    byte* tmpCert = NULL;
    word32 tmpCertSz;
    WC_DerCertList* tmpDerList;
    WC_DerCertList* cur;

    ret = wc_PKCS12_parse(pkcs12, (const char*)pass, &tmpKey, &tmpKeySz,
        &tmpCert, &tmpCertSz, &tmpDerList);

    if (ret == 0) {
        key = Napi::Buffer<uint8_t>::Copy(env, tmpKey, tmpKeySz);
        XFREE(tmpKey, NULL, DYNAMIC_TYPE_PUBLIC_KEY);

        cert = Napi::Buffer<uint8_t>::Copy(env, tmpCert, tmpCertSz);
        XFREE(tmpCert, NULL, DYNAMIC_TYPE_PKCS);

        cur = tmpDerList;

        while (cur != NULL) {
            derList[i] = Napi::Buffer<uint8_t>::Copy(env, cur->buffer,
                cur->bufferSz);
            cur = cur->next;
        }

        if (tmpDerList != NULL) {
            wc_FreeCertList(tmpDerList, NULL);
        }

        out.Set("key", key);
        out.Set("cert", cert);
    }

    return out;
}

Napi::Number bind_wc_d2i_PKCS12(const Napi::CallbackInfo& info)
{
    Napi::Env env = info.Env();
    int ret;
    uint8_t* der = info[0].As<Napi::Uint8Array>().Data();
    int derSz = info[1].As<Napi::Number>().Int32Value();
    WC_PKCS12* pkcs12 = info[2].As<Napi::External<WC_PKCS12>>().Data();

    ret = wc_d2i_PKCS12((const byte*)der, derSz, pkcs12);

    return Napi::Number::New(env, ret);
}

Napi::Buffer<uint8_t> nodejsPKCS12InternalToDer(const Napi::CallbackInfo& info)
{
    Napi::Env env = info.Env();
    WC_PKCS12* pkcs12 = info[0].As<Napi::External<WC_PKCS12>>().Data();
    int ret;
    Napi::Buffer<uint8_t> der;
    byte* tmpDer = NULL;
    int tmpDerSz = 0;

    ret = wc_i2d_PKCS12(pkcs12, &tmpDer, &tmpDerSz);

    if (ret > 0) {
        der = Napi::Buffer<uint8_t>::Copy(env, tmpDer, ret);
        XFREE(tmpDer, NULL, DYNAMIC_TYPE_PKCS);
    }

    return der;
}

void bind_wc_PKCS12_free(const Napi::CallbackInfo& info)
{
    WC_PKCS12* pkcs12 = info[0].As<Napi::External<WC_PKCS12>>().Data();

    wc_PKCS12_free(pkcs12);
}
